sk_dsp_comm/fec_conv.py

```diff
@@ -222,90 +272,211 @@ def viterbi_decoder(self,x,metric_type='three_bit'):
+                for m in range(self.Nstates):
+                    d1 = self.bm_calc(self.branches.bits1[m],
+                                    x[n:n+3],metric_type,
+                                    quant_level)
+                    d1 = d1 + cm_past[self.branches.states1[m]]
+                    d2 = self.bm_calc(self.branches.bits2[m],
+                                    x[n:n+3],metric_type,
+                                    quant_level)
+                    d2 = d2 + cm_past[self.branches.states2[m]]
+                    if d1 <= d2: # Find the survivor assuming minimum distance wins
+                        cm_present[m] = d1
+                        self.paths.traceback_states[m,:] = np.hstack((self.branches.states1[m],
+                                    tb_states_temp[int(self.branches.states1[m]),:]))
+                        self.paths.traceback_bits[m,:] = np.hstack((self.branches.input1[m],
+                                    tb_bits_temp[int(self.branches.states1[m]),:]))
+                    else:
+                        cm_present[m] = d2
+                        self.paths.traceback_states[m,:] = np.hstack((self.branches.states2[m],
+                                    tb_states_temp[int(self.branches.states2[m]),:]))
+                        self.paths.traceback_bits[m,:] = np.hstack((self.branches.input2[m],
+                                    tb_bits_temp[int(self.branches.states2[m]),:]))
+                # Update cumulative metric history
+                self.paths.cumulative_metric = np.hstack((cm_present, 
+                                            self.paths.cumulative_metric[:,:-1]))
+                
+                # Obtain estimate of input bit sequence from the oldest bit in 
+                # the traceback having the smallest (most likely) cumulative metric
+                min_metric = min(self.paths.cumulative_metric[:,0])
+                min_idx = np.where(self.paths.cumulative_metric[:,0] == min_metric)
+                if n >= 3*self.decision_depth-3:  # 3 since Rate = 1/3
+                    y[k] = self.paths.traceback_bits[min_idx[0][0],-1]
+                    k += 1
+            y = y[:k] # trim final length
         return y
 
-    def bm_calc(self,ref_code_bits, rec_code_bits, metric_type):
+    def bm_calc(self,ref_code_bits, rec_code_bits, metric_type, quant_level):
         """
         distance = bm_calc(ref_code_bits, rec_code_bits, metric_type)
         Branch metrics calculation
 
-        Mark Wickert February 2014
+        Mark Wickert and Andrew Smit October 2018
         """
 
-        if metric_type == 'three_bit': # squared distance metric
-            bits = binary(int(ref_code_bits),2)
-            ref_MSB = 7*int(bits[0],2)
-            ref_LSB = 7*int(bits[1],2)
-            distance = (rec_code_bits[0] - ref_MSB)**2
-            distance += (rec_code_bits[1] - ref_LSB)**2
+        if metric_type == 'soft': # squared distance metric

```

Comment:
Same comment as above. Perhaps look at using a Fractional and it's denominator here as well.

Reply:
Done